### PR TITLE
Fixes related to the use with MS Visual Studio Compiler

### PR DESCRIPTION
**Pull request by slavanap**
Additional overflow check (zero_copy_stream_impl.cc)

### CODE REVIEW

**Comment by slavanap** [comment]
I have signed CLA.

File: src/google/protobuf/compiler/cpp/cpp_message.cc [hunk 1]
```diff
@@ -1217,6 +1217,11 @@ GenerateDescriptorDeclarations(io::Printer* printer) {
       }
     }
 
+    printer->Print("#ifdef _MSC_VER" "\n"
+                   "  $name$OneofInstance() {}" "\n"
```
Comment (xfxyjwf) [inline src/google/protobuf/compiler/cpp/cpp_message.cc]: Why would VC require a constructor? I tested v3.0.0-alpha-1 on visual studio 2008 and 2013 but did not find such a problem.

File: src/google/protobuf/io/zero_copy_stream_impl.cc [hunk 2]
```diff
@@ -389,7 +390,8 @@ bool ConcatenatingInputStream::Skip(int count) {
     // to skip.
     int64 final_byte_count = streams_[0]->ByteCount();
     GOOGLE_DCHECK_LT(final_byte_count, target_byte_count);
-    count = target_byte_count - final_byte_count;
+	GOOGLE_DCHECK_LT(target_byte_count - final_byte_count, std::numeric_limits<int>::max());
```
Comment (xfxyjwf) [inline src/google/protobuf/io/zero_copy_stream_impl.cc]: Use spaces to indent.

File: src/google/protobuf/io/zero_copy_stream_impl.cc [hunk 3]
```diff
@@ -389,7 +390,8 @@ bool ConcatenatingInputStream::Skip(int count) {
...
+    count = (int)(target_byte_count - final_byte_count);
```
Comment (xfxyjwf) [inline src/google/protobuf/io/zero_copy_stream_impl.cc]: Use static_cast.

File: src/google/protobuf/stubs/template_util.h [hunk 4]
```diff
@@ -50,6 +50,14 @@
 #ifndef GOOGLE_PROTOBUF_TEMPLATE_UTIL_H_
 #define GOOGLE_PROTOBUF_TEMPLATE_UTIL_H_
 
+#if defined _MSC_VER
+#include <type_traits>
+#define __thread __declspec(thread)
+#endif
+#if defined IOS
+#define __thread // not supported on IOS devices
```
Comment (xfxyjwf) [inline src/google/protobuf/stubs/template_util.h]: Where is this used? This file does not seem like the place to define it.

File: src/google/protobuf/stubs/template_util.h [hunk 5]
```diff
@@ -73,14 +81,20 @@ struct identity_ {
...
 
-template<class T, T v>
-struct integral_constant {
-  static const T value = v;
-  typedef T value_type;
-  typedef integral_constant<T, v> type;
-};
-
-template <class T, T v> const T integral_constant<T, v>::value;
+#ifdef _MSC_VER
```
Comment (xfxyjwf) [inline src/google/protobuf/stubs/template_util.h]: Which VC version are you using?

**Comment by slavanap** [comment]
My bad that I used commit
	11/21/2014 03:18:53 SHA-1: 99aa0f9e8f1a88def7bdebf1385678559cda0707 *"Down-integrate from internal code base."* as a reference. 

This issue:
```
#define __thread __declspec(thread)
```
was already fixed by this definition @ common.h:317
```
#if defined(_MSC_VER)
#define GOOGLE_THREAD_LOCAL __declspec(thread)
#else
#define GOOGLE_THREAD_LOCAL __thread
#endif
```
in commit by you
	on 12/05/2014 SHA-1: d77877808fc9f00c3570353717d9659f8994ffc2 	*"Fix thread local annotatoin and add back type traits is_convertable for MSVC*", although it still does not valid for iOS build, because iOS does not support ```__thread``` attribute.

Another issue:
```
	#ifdef _MSC_VER
  		using std::tr1::integral_constant;
  		using std::tr1::is_convertible;
  		using std::tr1::is_enum;
	#else
	...
```
was fixed by you, too
	on 12/05/2014 SHA-1: be20ae0b6975071563ecc61f8372fd7936f174ed *"Fix compile issues and test failures in VS2008."*

The last issue (VS2008 error example below):
```
	error C2512: 'wl::protobuf::`anonymous-namespace'::valkeyOneofInstance' : no appropriate default constructor available
```
was fixed, but I still wondering why this was happening.
The one element in struct ```$name$OneofInstance``` that was needed a constructor is the one with type ```::google::protobuf::internal::ArenaStringPtr``` and it compiles well if we remove ```const``` before the member with this type. Here is the screenshots: [compiler error place](https://dl.dropboxusercontent.com/u/39984835/protobuf_errorplace.jpg) and [struct declaration](https://dl.dropboxusercontent.com/u/39984835/protobuf_buggyvariable.jpg).

So, my pull request became really short. Please excuse me for the inconvenience and not updating to the latest alpha release before commenting (too much work with lots of third-party libraries requires to think twice before updating to the latest release, that's why I had fixed issues with commit I checked out in November instead of keep tracking the development. Sorry.)

P.S. I'm using VS2008 (cl.exe version 15.00.30729.01) and VS2013.

**Pull request closed without merging**